plugins/inputs: New input for Wireguard server

LINKIWI opened the pull request:
This PR adds a new input for collecting metrics from the local [Wireguard](https://www.wireguard.com/) server.

High level changes:

- Add additional dependency to Gopkg: [wgctrl-go](https://github.com/WireGuard/wgctrl-go) (library for interacting with Wireguard implementation)
- Add new `plugins/inputs/wireguard` package with associated unit tests
- Update READMEs to reference the new plugin

Manual end-to-end verification:

```
# Happy path
$ sudo ./telegraf --config telegraf.conf --test
2019-09-08T20:20:22Z I! Starting Telegraf 
> wireguard_device,host=ONYX,name=tun0,type=linux_kernel firewall_mark=0i,listen_port=20029i 1567974023000000000
> wireguard_device,host=ONYX,name=tun0,type=linux_kernel peers=2i 1567974023000000000
> wireguard_peer,device=tun0,host=ONYX,public_key=NZTRIrv/ClTcQoNAnChEot+WL7OH7uEGQmx8oAN9rWE= allowed_ips=1i,persistent_keepalive_interval=0,protocol_version=1i 1567974023000000000
> wireguard_peer,device=tun0,host=ONYX,public_key=NZTRIrv/ClTcQoNAnChEot+WL7OH7uEGQmx8oAN9rWE= last_handshake_time=1567905087i,rx_bytes=261415128i,tx_bytes=334031704i 1567974023000000000
> wireguard_peer,device=tun0,host=ONYX,public_key=NZTRIrv/ClTcQoNAnChEot+WL7OH7uEGQmx8oAN9rWE= allowed_ips=1i,persistent_keepalive_interval=0,protocol_version=1i 1567974023000000000
> wireguard_peer,device=tun0,host=ONYX,public_key=NZTRIrv/ClTcQoNAnChEot+WL7OH7uEGQmx8oAN9rWE= last_handshake_time=1567973913i,rx_bytes=141607132i,tx_bytes=930069796i 1567974023000000000
```

```
# Plugin error during `Gather`
$ sudo ./telegraf --config telegraf.conf --test
2019-09-08T21:04:06Z I! Starting Telegraf 
2019-09-08T21:04:06Z E! [inputs.wireguard] Error in plugin: error enumerating Wireguard devices: err=file does not exist
2019-09-08T21:04:06Z E! [telegraf] Error running agent: One or more input plugins had an error
```

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

LINKIWI commented:
Would appreciate some help with the dependencies; this is my first time using `dep`. CI fails on `dep check` after I do this:

```
$ dep ensure -add golang.zx2c4.com/wireguard/wgctrl
$ dep check
# vendor is out of sync:
golang.zx2c4.com/wireguard/wgctrl: missing from vendor
```

even though `vendor/golang.zx2c4.com/wireguard/wgctrl` is definitely there.

LINKIWI commented on plugins/inputs/wireguard/wireguard.go:
```diff
@@ -0,0 +1,150 @@
+func (wg *Wireguard) Init() error {
+	var err error
+
+	if wg.initialized {
```
Cool, I didn't see `Init()` in the `telegraf.Input` interface so I assumed logic higher in the stack was unaware of input initialization routines.

LINKIWI commented on plugins/inputs/wireguard/README.md:
```diff
@@ -0,0 +1,63 @@
+By default, Telegraf runs as the `telegraf` system user. Wireguard
+implementations that run in kernelspace (as opposed to userspace) require
+userspace programs to run as root in order to communicate with the module.
+Either update the system udev rules for the `telegraf` user or run Telegraf as
```
The advice I've offered here is actually incorrect. Seems like I need to either run Telegraf as root or add the `CAP_NET_ADMIN` capability to the Telgraf binary. I'll document this in greater detail.

LINKIWI commented on Gopkg.lock:
```diff
@@ -828,6 +828,38 @@
+[[projects]]
+  branch = "master"
+  digest = "1:d9ee5e12190ff796cadfad8a9aae2ad57a303b1d4d476518ace3a5dc01ae4c3a"
+  name = "github.com/mdlayher/netlink"
```
This is a transitive dependency introduced by the addition of the direct dependency github.com/mdlayher/wireguardctrl. Do you mean you would prefer re-implementing a portion of github.com/mdlayher/wireguardctrl using the existing netlink library, and avoid introducing additional dependencies entirely? Or do you mean you would prefer using a fork of github.com/mdlayher/wireguardctrl that uses vishvananda/netlink so as to not introduce a new transitive netlink dependency?

LINKIWI replied:
(I would probably prefer the latter, since I'm planning on forking mdlayher/wireguardctrl anyways for use in a separate future project, but happy to try whichever approach unblocks merging here)

danielnelson replied:
The most ideal would be the former, since it would result in the smallest overall code size, but the either way would be acceptable.

It looks like there are 2 methods for interfacing with wireguard: netlink or a userspace unix socket/named pipe.  On Linux, is the unix socket created even when using the wireguard interface, or is it only created when using the userspace implementation?  If the socket was always created that could be the ideal way for us to monitor it.

LINKIWI replied:
A Wireguard Unix socket is only created in userspace implementations (and even then it is just an I/O convention, since it's not explicitly laid out in the spec as far as I am aware). It seems like netlink is the only way to communicate with Wireguard in its kernelspace implementation.

I didn't look very closely into it, but I believe the library dependency I've added chooses automatically based on the conventions from the first-party userspace Wireguard implementation.

LINKIWI commented:
Hi all, thanks for reviewing this new input.

@danielnelson I've taken a look at the feasibility of substituting vishvananda/netlink for the transitive dependency mdlayher/netlink introduced by the direct dependency mdlayher/wireguardctrl. This seems like a significantly non-trivial amount of work, especially since the package has a nested dependency on one of the author's other netlink libraries. From grokking through the author's source code, this effort would probably involve adding almost a thousand lines of Wireguard/netlink-related library code to Telegraf, which is probably not very desirable.

Since this patch is purely for my personal use, I've opted to just run my fork of Telegraf (with the additional dependency). I'm happy to relinquish this PR to the community in case anyone else has the time to pick up the additional refactoring needed to get this merged.

danielnelson commented:
Fair enough, let's keep this PR open for now for visibility until we can integrate it or similar functionality.

danielnelson commented:
@LINKIWI After some more consideration, I'd like to move forward to include this PR. https://github.com/influxdata/telegraf/pull/6587#issuecomment-550121162

Might take me a bit more to get back to this but I'm requesting a new review from myself.

danielnelson commented on docs/LICENSE_OF_DEPENDENCIES.md:
```diff
@@ -72,6 +72,7 @@ following works:
 - github.com/mailru/easyjson [MIT License](https://github.com/mailru/easyjson/blob/master/LICENSE)
 - github.com/matttproud/golang_protobuf_extensions [Apache License 2.0](https://github.com/matttproud/golang_protobuf_extensions/blob/master/LICENSE)
 - github.com/mdlayher/apcupsd [MIT License](https://github.com/mdlayher/apcupsd/blob/master/LICENSE.md)
+- github.com/mdlayher/wireguardctrl [MIT License](https://github.com/mdlayher/wireguardctrl/blob/master/LICENSE.md)
```
Need to add licenses for all new repos, this doesn't necessarily need to be done in this PR, we can do it at merge time.

LINKIWI replied:
Got it, I've unversioned changes to this file so that it can be done at merge time.

LINKIWI commented:
Thanks for the review. I'll try to update this patch some this week (haven't touched this code in quite some time).

LINKIWI commented:
Rebased against master and addressed comments. PTAL, thanks.

A couple notes:
* `go mod tidy` updated some unrelated dependencies, the changes from which I've checked in.
* Seems like the example `telegraf.conf` was pretty out of date; `telegraf config > etc/telegraf.conf`generated a lot of unrelated changes. I've selectively checked in only the changes related to this PR.
